service/history/callbacks/statemachine.go

```diff
@@ -0,0 +1,180 @@
+// The MIT License
+//
+// Copyright (c) 2024 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package callbacks
+
+import (
+	"fmt"
+	"net/url"
+	"time"
+
+	commonpb "go.temporal.io/api/common/v1"
+	enumspb "go.temporal.io/api/enums/v1"
+	failurepb "go.temporal.io/api/failure/v1"
+	persistencespb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common/backoff"
+	"go.temporal.io/server/service/history/statemachines"
+	"go.temporal.io/server/service/history/tasks"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type adapter struct{}
+
+func (adapter) GetState(data *persistencespb.CallbackInfo) enumspb.CallbackState {
+	return data.PublicInfo.State
+}
+
+func (adapter) SetState(data *persistencespb.CallbackInfo, state enumspb.CallbackState) {
+	data.PublicInfo.State = state
+}
+
+func (adapter) OnTransition(data *persistencespb.CallbackInfo, from, to enumspb.CallbackState, env statemachines.Environment) {
+	// TODO: consider moving this into the "framework".
+	data.Version = env.GetVersion()
+	if from == enumspb.CALLBACK_STATE_SCHEDULED {
+		// Reset all of previous attempt's information.
+		data.PublicInfo.Attempt++
+		data.PublicInfo.LastAttemptCompleteTime = timestamppb.New(env.GetCurrentTime())
+		data.PublicInfo.LastAttemptFailure = nil
```

Comment:
This means if callback completed after a few retries, user won't know what's the previous failure that cause the backoff & retry? 
I think we have this information available for activities (on activity started event).

Reply:
Yeah, this is what I meant here. I even have a test that verifies it.

I considered keeping the last failed attempt here but was worried that it would confuse users if they see the failure even though the callback ended up successful.

I agree there could be value in the added visibility though.

Reply:
@MichaelSnowden you may want to weigh in here too.

I might merge once CI passes but we can revisit this later.